### DISCUSSION TITLE
Discussion: Event audience

### DISCUSSION THREAD

**Discussion by samthorold** [ki-oss/hades#18] [Ideas]
Hi friends 👋, starting a discussion here re how the idea of an event audience might be introduced to `hades`.

The `Event` [docs](https://ki-oss.github.io/hades/api_reference/event/) talk about this briefly.

> Another thing which may be useful is some kind of audience identifier, which can help processes distinguish quickly whether to do anything with a give event.

Before discussing potential implementations, I reckon it would behoove us to be mostly on the same page re aims, existing approaches, and problems with the existing approaches.

- [ ] Happy that the motivating example covers the "problem"?
- [ ] Happy we've captured the existing approaches?
- [ ] Happy we've captured the problems with the existing approaches?

### Motivating example

Suppose you want to simulate searching for a flatmate as someone with a room to rent.
In the most basic example, you might have a single `Resident` process looking for a flatmate and three `ProspectiveFlatmate` processes looking for a room.

1. Initially, an event, `RoomAvailable`, could be broadcast to all processes.
2. The `Resident` processes may respond with an `ApplicationMade` event.
3. The `Resident` process may respond to two of the `ProspectiveFlatmate`s that they would like to meet them for a drink to get to know them better, a `SecondChatScheduled` event perhaps.
4. After some period of time, say, a `FinalDecisionMade` event is broadcast to the two `ProspectiveFlatmate`s who were met for a drink with the final selection.

**How to send `SecondChatScheduled` and `FinalDecisionMade` events only to the relevant `ProspectiveFlatmate`s?**

_There are a number of areas we could make this more complex e.g. multiple `Resident`s. Could argue the `FinalDecisionMade` should be sent to all `Resident`s so they know the room is gone._

### Existing approach(es)

The `SecondChatScheduled` and `FinalDecisionMade` events could include a list of `ProspectiveFlatmate` identifiers that they pertain to.

The `ProspectiveFlatmate` not included in the list of identifiers could ignore those events.

### Problems with the existing approach(es)

- More events being broadcast than is strictly necessary.
- Potential information "leakage" with processes able to learn information they would not have in the real world system being modelled.
    - Difficult to have "private" sub-networks of processes that communicate with the same events.
- On users to implement their own way of dealing with this.

---

Propose that I amend this post as comments come in.

Probably one for a different post, benefits to enabling the discussion feature on `ki-oss` / `hades`?

**Comment by Rested** [comment]
Think this would be a great addition! 

I think the motivating example does cover a problem and capture the existing approaches. 

I think one benefit of the existing approach is that broadcast by default simplifies things when picking up the framework which I'd be keen to preserve therefore I think an implementation should be non-breaking/opt-in.

One thing to consider would be if it is on process type or instance too. The motivating example could be changed to having the `RoomAvailable` event only broadcast to all `Resident` processes.
We may want to offer either.

Maybe something like a `ProcessScopedEvent` which specifies permissable processes ahead of time in the event definition and then like an `InstanceScopedEvent` for the instance specific ones dynamically?

**Reply by samthorold** [reply]
Definitely helpful to be opt-in 👍

It isn't clear to me what the practical differences are between `ProcessScopedEvent` and NACKing in the usual way with `case: _`.

I'll take a swing at a working example ⌨️

**Reply by Rested** [reply]
one thing would be speed - hades would not create a coroutine for sending those events to the out of scope processes (and so be a bit faster). Other difference would be that processes can't choose to ACK if they want to with a `case ProcessScopedEvent`

**Comment by Rested** [comment]
+1 for enabling the discussion feature - will do it!

**Reply by reuben-ki** [reply]
Converted now @samthorold!
